File: game.lua
Game = {
    map,
    mapWidth,
    mapHeight,
    flagLeft, 
    mineFound,
    mine,
    state 
}

local function shuffleTab(a)
	local j = 0;
	local valI = nil
	local valJ = nil
	local l = table.getn(a)-1;
	while (l > -1) do
		j = 1+math.floor(math.random() * l);
		valI = a[l];
		valJ = a[j];
		a[l] = valJ;
		a[j] = valI;
		l = l - 1;
	end
	return a;
end

function Game.initModel(w,h,m)
	Game.mapWidth = w
	Game.mapHeight = h
	Game.flagLeft = m
	Game.mineFound = 0
	Game.mine = m
	Game.state = "playing"

	Game.map = {}
	for x=1,Game.mapWidth do
		Game.map[x] = {}
		for y=1,Game.mapHeight do
			Game.map[x][y] = {model_mineAround= 0, model_discovered= false, model_flaged=false}
		end
	end

	local place = {}
	for i=0,Game.mapWidth*Game.mapHeight do
		place[i]=i
	end
	shuffleTab(place);
	for i=1,Game.mine do
		local x= place[i]%Game.mapWidth;
		local y= math.floor(place[i]/Game.mapWidth);
		Game.placeMine(x,y)
	end
end

function Game.placeMine(x,y) 	
	if(x>=1 and y>=1 and x<=Game.mapWidth and y<=Game.mapHeight) then

		if(Game.map[x][y].model_mineAround == 9) then
			return false
		end

		Game.map[x][y].model_mineAround=9;
		Game.neighbourMinePlaced(x-1,y-1);
		Game.neighbourMinePlaced(x  ,y-1);
		Game.neighbourMinePlaced(x+1,y-1);
		Game.neighbourMinePlaced(x-1,y  );
		Game.neighbourMinePlaced(x+1,y  );
		Game.neighbourMinePlaced(x-1,y+1);
		Game.neighbourMinePlaced(x  ,y+1);
		Game.neighbourMinePlaced(x+1,y+1);

		return true
	end
	return false
end

function Game.neighbourMinePlaced(x,y)
	if(x>=1 and y>=1 and x<=Game.mapWidth and y<=Game.mapHeight) then
		if Game.map[x][y].model_mineAround == 9 then
			return 
		end 

		Game.map[x][y].model_mineAround = Game.map[x][y].model_mineAround+1

	end

end

function Game.flag(x,y)
	local ret = false
	if(x>=1 and y>=1 and x<=Game.mapWidth and y<=Game.mapHeight) then
		local gameBox = Game.map[x][y]
		if(	gameBox.model_discovered == false ) then
			if(Game.flagLeft == 0 and gameBox.model_flaged == false) then return ret end
			ret = true
			if gameBox.model_flaged then
				Game.flagLeft = Game.flagLeft+1
			else 
				Game.flagLeft = Game.flagLeft-1
			end		

			gameBox.model_flaged = not gameBox.model_flaged;

			if(gameBox.model_mineAround == 9) then

				if gameBox.model_flaged then
					Game.mineFound = Game.mineFound+1
				else 
					Game.mineFound = Game.mineFound-1
				end		

				if(Game.mineFound == Game.mine) then 
					Game.state = "win"
				end	
			end
		end
	end
end


function Game.reveal(x,y)
	local ret = false
	if(x>=1 and y>=1 and x<=Game.mapWidth and y<=Game.mapHeight) then

		local gameBox = Game.map[x][y]
		if 	gameBox.model_discovered == false  then
			if gameBox.model_flaged then return ret end
			gameBox.model_discovered = true
			ret = true

			if(gameBox.model_mineAround == 9) then
				gameBox.model_discovered = true
				Game.state = "loose"
			elseif(gameBox.model_mineAround == 0) then
				revealAround(x,y);
			end
		end
	end

	return ret;
end

function revealAround(x,y)
	if(x>=1 and y>=1 and x<=Game.mapWidth and y<=Game.mapHeight) then
		Game.reveal(x-1,y-1);
		Game.reveal(x  ,y-1);
		Game.reveal(x+1,y-1);
		Game.reveal(x-1,y  );
		Game.reveal(x+1,y  );
		Game.reveal(x-1,y+1);
		Game.reveal(x  ,y+1);
		Game.reveal(x+1,y+1);
	end
end


function Game.tileFromModel(model)
	local id = model.model_mineAround

	if Game.state == "playing" then

		if model.model_discovered then            
			if id == 0 then
				return 6
			elseif id == 9 then 
				return  3
			else
				return id+6
			end
		elseif model.model_flaged then
			return 0
		else
			return 5
		end

	else
		if not model.model_flaged then            
			if id == 0 then
				return 6
			elseif id == 9 then 
				if model.model_discovered then
					return  3
				else 
					return 2
				end
			else
				return id+6
			end
		elseif id == 9 then
			return 0
		else
			return 4
		end


	end
end 

File: src/game.ts


function shuffleTab<T>(array:Array<T>): Array<T>{
	let l:number = array.length;

	while(l>=0){
		let j = 1+math.floor(math.random() * l);
		let valI:T = array[l];
		let valJ:T = array[j];
		array[l] = valJ;
		array[j] = valI;
		l = l - 1;
	}
	return array;
}

export type State = "playing" | "win" | "loose"

class Cell{
	mineAround:number;
	discovered:boolean;
	flagged:boolean;

	constructor(m:number,d:boolean,f:boolean){
		this.mineAround=m;
		this.discovered=d;
		this.flagged=f;
	}
}

export class Game{
	width:number;
	height:number;
	mines:number;
	minesFound:number;
	flagLeft:number;
	state:State;
	map:Array<Array<Cell>>;

	constructor(w:number,h:number,m:number){
		this.width = w;
		this.height = h;
		this.mines = m;
		this.minesFound = 0; 
		this.flagLeft = m;
		this.state = "playing";

		this.map = [];
		for(let x=0;x<this.width;x++){
			this.map[x] = [];
			for(let y=0;y<this.height;y++){
				this.map[x][y] = new Cell(0,false,false);
			}
		}
	
		let place:Array<number>= [];
		for (let i=0;i<this.width*this.height;i++){
			place[i]=i
		}
		place=shuffleTab(place);

		for (let i=0;i<this.mines;i++){
			let x:number = place[i]%this.width;
			let y:number = math.floor(place[i]/this.width);
			this.placeMine(x,y);
		}
	}



	placeMine(x:number,y:number):boolean {
		if(x<0){ return false; }
		if(y<0){ return false; }
		if(x>=this.width){ return false; }
		if(y>=this.height){ return false; }

		if(this.map[x][y].mineAround == 9){ return false }

		this.map[x][y].mineAround = 9;
		this.neighbourMinePlaced(x-1,y-1);
		this.neighbourMinePlaced(x  ,y-1);
		this.neighbourMinePlaced(x+1,y-1);
		this.neighbourMinePlaced(x-1,y  );
		this.neighbourMinePlaced(x+1,y  );
		this.neighbourMinePlaced(x-1,y+1);
		this.neighbourMinePlaced(x  ,y+1);
		this.neighbourMinePlaced(x+1,y+1);

		return true;
	}

	neighbourMinePlaced(x:number,y:number){
		if(x<0){ return }
		if(y<0){ return  }
		if(x>=this.width){ return }
		if(y>=this.height){ return }

		if(this.map[x][y].mineAround == 9){ return }

		this.map[x][y].mineAround++;
	}

	flag(x:number,y:number):boolean{
		if(x<0){ return false }
		if(y<0){ return  false }
		if(x>=this.width){ return false }
		if(y>=this.height){ return false }
		
		let cell:Cell = this.map[x][y];
		
		if(cell.discovered ) { return false }
		if (this.flagLeft == 0 && cell.flagged == false){ return false }

		this.flagLeft += cell.flagged?-1:1;
		cell.flagged = !cell.flagged;

		if (cell.mineAround != 9 ) { return true }

		this.minesFound += cell.flagged?-1:1;
		if(this.minesFound ==this.mines){
			this.state = "win";
		}
		return true;
	}

	reveal(x:number,y:number) :boolean{
		if(x<0){ return false }
		if(y<0){ return  false }
		if(x>=this.width){ return false }
		if(y>=this.height){ return false }
		
		let cell:Cell = this.map[x][y];

		if(cell.discovered ) { return false }
		if(cell.flagged ) { return false }
	
		cell.discovered = true;
		if (cell.mineAround == 9 ){
			this.state = "loose";
			return true;
		}
		if (cell.mineAround == 0 ){
			this.reveal(x,y);
		}

		return true;
	}

	revealAround(x:number,y:number) {
		if(x<0){ return }
		if(y<0){ return }
		if(x>=this.width){ return }
		if(y>=this.height){ return }

		this.reveal(x-1,y-1);
		this.reveal(x  ,y-1);
		this.reveal(x+1,y-1);
		this.reveal(x-1,y  );
		this.reveal(x+1,y  );
		this.reveal(x-1,y+1);
		this.reveal(x  ,y+1);
		this.reveal(x+1,y+1);
	}

	tileAt(x:number,y:number):number{
		if(x<0){ return -1 }
		if(y<0){ return  -1 }
		if(x>=this.width){ return -1 }
		if(y>=this.height){ return -1 }
		
		let c:Cell = this.map[x][y];

		if (this.state == "playing")
		{
			if(c.discovered){
				switch (c.mineAround) {
					case 0: return 6;
					case 9: return 3;
					default: return c.mineAround + 6;
				}
			}
			if(c.flagged){
				return 0;
			}
			return 5;

		} 
		if(!c.flagged){
			switch (c.mineAround) {
				case 0: return 6;
				case 9: return c.discovered?3:2;
				default: return c.mineAround + 6;
			}
		}
		if(c.mineAround ==9){
			return 0;
		}
		return 4;
	}

	tiles():Array<Array<number>>{

		let tiles:Array<Array<number>>=[];
		for(let x=0;x<this.width;x++){
			tiles[x] =[];
			for(let y=0;y<this.height;y++){
				tiles[x][y] = this.tileAt(x,y);
			}
		}
		return tiles;
	}

}



